src/Hocon.Configuration/Config.cs

```diff
@@ -18,82 +20,54 @@ namespace Hocon
+        public new static Config Empty { get; } = new Config();
 
-        static Config()
+        public static Config Create(HoconElement root)
         {
-            EmptyValue = new HoconValue(null);
-            EmptyValue.Add(new HoconObject(EmptyValue));
+            return new Config(root);
         }
 
-        [Obsolete("For json serialization/deserialization only", true)]
-        protected Config()
+        private Config()
         {
+            Root = this;
         }
 
         /// <inheritdoc />
         /// <summary>
         ///     Initializes a new instance of the <see cref="Config" /> class.
         /// </summary>
-        protected Config(HoconValue value)
-        {
-            Value = (HoconValue)value.Clone(null);
-            Root = (HoconValue)value.Clone(null);
-        }
-
-        /// <inheritdoc cref="Config(HoconValue)" />
-        protected Config(HoconValue value, Config fallback) : this(value)
-        {
-            MergeConfig(fallback);
-        }
-
-        /// <inheritdoc cref="Config(HoconValue)" />
-        /// <param name="root">The root node to base this configuration.</param>
-        /// <exception cref="T:System.ArgumentNullException">"The root value cannot be null."</exception>
-        public Config(HoconRoot root)
+        public Config(HoconElement root):base(root)
         {
-            Value = (HoconValue)root.Value.Clone(null);
-            Root = (HoconValue)root.Value.Clone(null);
-
-            if (!(root is Config cfg))
-                return;
-
-            foreach(var value in cfg._fallbacks)
+            if(root is Config cfg)
+            {
+                Root = cfg.Root;
+                _fallbacks = cfg._fallbacks.ToList();
+                _cache = cfg._cache;
+            }  
+            else 
             {
-                InsertFallbackValue(value);
+                Root = this;
             }
         }
 
-        /// <inheritdoc cref="Config(HoconValue)" />
-        /// <param name="root">The configuration to use as the primary source.</param>
-        /// <param name="fallback">The configuration to use as a secondary source.</param>
-        /// <exception cref="ArgumentNullException">The source configuration cannot be null.</exception>
-        public Config(HoconRoot root, Config fallback) : this(root)
-        {
-            MergeConfig(fallback);
-        }
-
-        /// <summary>
-        ///     Identical to <see cref="HoconConfigurationFactory.Empty" />.
-        /// </summary>
-        /// <remarks>
-        ///     Added for brevity and API backwards-compatibility with Akka.Hocon.
-        /// </remarks>
-        public static Config Empty => CreateEmpty();
-
-        protected List<HoconValue> _fallbacks { get; } = new List<HoconValue>();
-        public virtual IReadOnlyList<HoconValue> Fallbacks => _fallbacks.ToList().AsReadOnly();
+        private ConcurrentDictionary<string, HoconElement> _cache = new ConcurrentDictionary<string, HoconElement>();
```

Comment:
We're going to need to get rid of the cache. Memory consumption increased 6x on this PR.

## Before

# Hocon.Tests.Performance.ConfigLookupSpec+Config_Lookup_HoconPath_10_Deep_throughput
__Tests how quickly Config can perform a 10 deep path lookup using a HoconPath path__
_2/25/2020 5:45:12 PM_
### System Info
```ini
NBench=NBench, Version=2.0.1.0, Culture=neutral, PublicKeyToken=null
OS=Microsoft Windows NT 6.2.9200.0
ProcessorCount=2
CLR=4.0.30319.42000,IsMono=False,MaxGcGeneration=2
```

### NBench Settings
```ini
RunMode=Throughput, TestMode=Measurement
NumberOfIterations=13, MaximumRunTime=00:00:01
Concurrent=True
Tracing=True
```

## Data
-------------------

### Totals
|          Metric |           Units |             Max |         Average |             Min |          StdDev |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|TotalCollections [Gen0] |     collections |          138.00 |          138.00 |          138.00 |            0.00 |
|TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |           bytes |    1,227,752.00 |    1,227,752.00 |    1,227,752.00 |            0.00 |
|[Counter] LookupOp |      operations |       92,086.00 |       92,086.00 |       92,086.00 |            0.00 |

### Per-second Totals
|          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|TotalCollections [Gen0] |     collections |          140.30 |          134.00 |          127.71 |            3.78 |
|TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |           bytes |    1,248,193.67 |    1,192,135.24 |    1,136,214.79 |       33,612.93 |
|[Counter] LookupOp |      operations |       93,619.20 |       89,414.61 |       85,220.37 |        2,521.10 |

## After

# Hocon.Tests.Performance.ConfigLookupSpec+Config_Lookup_HoconPath_10_Deep_throughput
__Tests how quickly Config can perform a 10 deep path lookup using a HoconPath path__
_2/26/2020 8:20:18 PM_
### System Info
```ini
NBench=NBench, Version=2.0.1.0, Culture=neutral, PublicKeyToken=null
OS=Microsoft Windows NT 6.2.9200.0
ProcessorCount=2
CLR=4.0.30319.42000,IsMono=False,MaxGcGeneration=2
```

### NBench Settings
```ini
RunMode=Throughput, TestMode=Measurement
NumberOfIterations=13, MaximumRunTime=00:00:01
Concurrent=True
Tracing=True
```

## Data
-------------------

### Totals
|          Metric |           Units |             Max |         Average |             Min |          StdDev |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|TotalCollections [Gen0] |     collections |           33.00 |           33.00 |           33.00 |            0.00 |
|TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |           bytes |    6,054,296.00 |    6,054,296.00 |    6,054,296.00 |            0.00 |
|[Counter] LookupOp |      operations |      209,371.00 |      209,371.00 |      209,371.00 |            0.00 |

### Per-second Totals
|          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
|---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
|TotalCollections [Gen0] |     collections |           34.26 |           33.36 |           32.58 |            0.62 |
|TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
|TotalBytesAllocated |           bytes |    6,285,038.62 |    6,120,678.42 |    5,977,090.52 |      113,842.84 |
|[Counter] LookupOp |      operations |      217,350.59 |      211,666.65 |      206,701.06 |        3,936.94 |


The performance improvement is nice - but we read from HOCON values fairly infrequently, whereas they are retained in memory for the duration of the application and we need to be gentle with memory consumption considering that we will run inside environments like IOT where RAM is at a premium. If we want to add caching to improve performance in a specific area of HOCON, that should be up to the users sub-classing `Config` - not something `Config` does carte blanche.

Reply:
I need to rework the perf spec so that it check the memory allocated with and without the config creation step.